### ISSUE TITLE
Rationalize \e in macros / Queue clearing

### ISSUE THREAD

**Issue by flambard-took** [mangband/mangband#1208]
In both 1.4.0 and 1.5.0, pressing ESCAPE sends a "clear the command queue" packet to the server.

There is also a hack, which ignores subsequent escapes when a macro is processed.

That means:
`\eq1\eq1\eq1` - sends 1 clear queue packet and 3 quaff packets. (nice)
`\eq1q1q1` - does the same thing. (nice)
`\eq1` pressed 3 times - sends 1 clear queue packet and 1 quaff packet - 3 times. (evil)

Thus, when you spam the server with commands, you're actually clearing the queue every time, and NOT queuing any commands.

Consider this:
`\er1`, `\eq1` - to read phase door and quaff a heal - will actually ONLY quaff a potion.
`\er1\eq1` (as a single macro) - will work as expected.

So, there *is* a way to do what you want, but I see many problems with this:
- players are *not* educated to write such macros
- players *are* educated to start each macro with \e "to exit any pending prompts"
- it's unfeasible to prepare all such macro combinations, I *feel*, that players would expect 2 macro presses to work as 2 normal command presses. 

I've also seen a `\e\e\e\e` pattern, which is a recommended way to "definitely exit any pending prompts" (we actually have this in our .prf files, binded to arrow keys and numpad). I think, that such a macro *should* clear the command queue.

**On other other hand**, players are *kinda* used to spam `\ef1*t` for firing missiles, and *kinda* trained to just hold the macro key down (or mash it rapidly), *kinda* expecting the queue to be cleared. (Same applies to offensive magic)

What I mean here, is that the command-clearing effect is **felt** in such cases, and it's more or less fine.

My question is: are my assumptions correct? Or are players perfectly aware of this weirdness and use it to their benefit? Is `\er1`, `\eq1` only executing last action a normal and expected behavior? Because if it is, then we're good and no change is required (maybe better documentation, and that's it). But if it's NOT expected behavior, we should definitely fix it.

If a fix is needed I see the following options:
- `\e\e` to clear queue, `\e` to escape prompts.
- re-educate players to write macros as `q1`,`r1`, without the `\e` part (but that will hinder the "escape any prompt" part)
- add a separate command for queue clearing (Ctrl+C, for example), and a special escape code `\C` (for example), so players can have very precise control over the queue mechanism
- disable queue clearing in macros **completely**, so that pressing actual Escape on keyboard works, but `\e` in macros doesn't

In any case, I feel that there at least *some* players, who are dying because of this, either due to documentation being too vague either due to this mechanism being completely broken.

What **I**, personally, want, is a macro that doesn't clear the queue, yet escapes command prompts, and that's impossible to achieve with the current system.

**Comment by flambard-took** [comment]
> My question is: are my assumptions correct? Or are players perfectly aware of this weirdness and use it to their benefit? Is \er1, \eq1 only executing last action a normal and expected behavior? 

https://mangband.org/forum/viewtopic.php?f=4&t=2206&p=9944 according to Zal, my assumptions were **not** correct, and players **are** aware of this and **do** use the system properly.

So, I'm guessing no change is needed after all. I'll wait for more opinion and then close this ticket, if that's the case.

**Comment by billsey** [comment]
I've always wondered if we can get into a race situation, where someone mashing the macro key for \ef1*t manages to abort the previous command before it gets executed. You'd end up firing once and aborting every other attempt. On the other hand you could also end up filling the buffer so full of fire commands that when you changed to the macro for quaff healing it'd take several seconds to process. Either scene could be bad. I think I'd like the behavior to change to allow for a double (or more) escape sequence clearing the command queue and a single escape to just bypass prompts.

**Comment by flambard-took** [comment]
>I've always wondered if we can get into a race situation, where someone mashing the macro key for \ef1*t manages to abort the previous command before it gets executed.

Yes, it's totally possible, and in fact happens. I wasn't aware of this myself...

> I think I'd like the behavior to change to allow for a double (or more) escape sequence clearing the command queue and a single escape to just bypass prompts.

I kinda like that too, I'm also thinking something like `\n\e` (if we can even use `\n`, macro processing is weird) to "keep the queue, and escape prompts" and regular `\e` to "clear queue AND escape the prompts" is an OK approach, would not break existing behavior, yet allow to do what we want.

**Comment by flambard-took** [comment]
https://www.tomenet.eu/guide.php?chapter=3.9
>)               clears the keyboard buffer (all previous key presses that
                  still have to be processed will be forgotten).
> *I recommend to have all macros start with the '\e)' sequence.*

(I'm assuming by "keyboard buffer" they actually mean server-side command queue, but I'm not sure)
